### PR TITLE
Add the new 'maintenance' privilege containing 4 actions (#29998)

### PR DESCRIPTION
**Pull request by amirhmd**
- Have you signed the [contributor license agreement](https://www.elastic.co/contributor-agreement)?
- Have you followed the [contributor guidelines](https://github.com/elastic/elasticsearch/blob/master/CONTRIBUTING.md)?
- If submitting code, have you built your formula locally prior to submission with `gradle check`?
- If submitting code, is your pull request against master? Unless there is a good reason otherwise, we prefer pull requests against master and will backport as needed.
- If submitting code, have you checked that your submission is for an [OS and architecture that we support](https://www.elastic.co/support/matrix#show_os)?
- If you are submitting this code for a class then read our [policy](https://github.com/elastic/elasticsearch/blob/master/CONTRIBUTING.md#contributing-as-part-of-a-class) for that.

### CODE REVIEW

**Comment by tvernum** [comment]
@amirhmd One of your "fix merge conflict" commits has introduced uninentional changes.

Can you please have a look at the diff, and have a go at cleaning it up. Let us know if you need help - we may need to perform a little bit of git surgery.

**Comment by amirhmd** [comment]
hi @tvernum 
could you please confirm if it sounds good now?

**Comment by amirhmd** [comment]
`f7933f65fd1 HEAD@{2}: commit (merge): Fix merge conflict`
`363f1ff89b7 HEAD@{3}: commit: Add the new 'maintenance' privilege containing 4 actions (#29998)`

`git reset --hard  f7933f65fd1`
`HEAD is now at f7933f65fd1 Fix merge conflict`

`git push origin master -f [it doesnt allow me to push to 'origin' because my current branch was behind]`
`Total 0 (delta 0), reused 0 (delta 0)`
`To https://github.com/amirhmd/elasticsearch.git`
`+cb39d77a4ec...f7933f65fd1 master -> master (forced update)`

**Comment by tvernum** [comment]
Thanks @amirhmd the diff looks good.
However, it still shows a merge conflit in GitHub, so I think there's still some work to do before we get to merging this.

**Comment by tvernum** [comment]
@elasticmachine ok to test.

**Comment by amirhmd** [comment]
hi @albertzaharovits one test is failing as I was explained in the comment. 
`IndexPrivilegeTests/testUser15/`
The other one which is failing `XPackRestIT.test {p0=privileges/11_builtin/Test get builtin privileges}` is because there is 33 Cluster Priviledge, I didnt increase this one in `11_builtin` but at the time of commit it shows a conflict and I forced to change it. 
could you please advise on this?

File: x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/security/authz/privilege/IndexPrivilege.java
Comment (albertzaharovits) [inline x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/security/authz/privilege/IndexPrivilege.java]: The test that it's failing `{"type":"security_exception","reason":"action [indices:admin/refresh[s]] is unauthorized for user [u15]"}}]}}` hints that `indices:admin/refresh[s]` is unauthorized.
The refresh action translates into shard wise sub-actions, which are also authorized by themselves.
You need to add replace `indices:admin/refresh` with `indices:admin/refresh*`.

**Comment by albertzaharovits** [comment]
I left a comment to hint at why the test is failing, I'll take a closer look in the morning.

File: x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/security/authz/privilege/IndexPrivilege.java
Comment (albertzaharovits) [reply x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/security/authz/privilege/IndexPrivilege.java]: This should be:
```
private static final Automaton MAINTENANCE_AUTOMATON = patterns("indices:admin/refresh*", "indices:admin/flush*", "indices:admin/synced_flush", "indices:admin/forcemerge*");
```
The motive is that, as mentioned above, a cluster action might spawn other actions, both on the node first handling the request as well as other nodes on the cluster. Every such action is authorized individually, i.e. the authorization status of the parent action is not consulted. For example, for the `refresh` (and `flush`) action translate to `indices:admin/refresh[s]` (and `indices:admin/flush[s]`) actions for every index shard, which are then translated into `indices:admin/refresh[p]` and `indices:admin/refresh[r]` for the primary and replica shards. The `maintenance` privilege should grant all these action types, and we customary add the wildcard star to the name of the originating action instead of iterating all the sub-action names.

File: x-pack/docs/en/security/authorization/privileges.asciidoc
Comment (albertzaharovits) [inline x-pack/docs/en/security/authorization/privileges.asciidoc]:
```Suggestion
Permits refresh, flush, synced flush and force merge operations. No privilege to read or write index data or otherwise manage the index.
```

File: x-pack/plugin/src/test/resources/rest-api-spec/test/privileges/11_builtin.yml
Comment (albertzaharovits) [inline x-pack/plugin/src/test/resources/rest-api-spec/test/privileges/11_builtin.yml]: You're adding a new _index_ privilege not a _cluster_ one, so I believe this should instead be:
```suggestion
   - length: { "index" : 18 }
```

File: x-pack/plugin/security/src/test/java/org/elasticsearch/integration/IndexPrivilegeTests.java
Comment (albertzaharovits) [inline x-pack/plugin/security/src/test/java/org/elasticsearch/integration/IndexPrivilegeTests.java]: I don't think we need this fluff. This is testing the authorization of composite requests but refresh and flush are not part of those composite type of requests.
In addition to `assertUserIsAllowed("u15", "maintenance", "a");` I would also check that `crud` is not allowed, and we should also test that other users in this class don't have the `maintenance` priv (right now it only checks the positive case).

File: x-pack/plugin/security/src/test/java/org/elasticsearch/integration/IndexPrivilegeTests.java
Comment (albertzaharovits) [inline x-pack/plugin/security/src/test/java/org/elasticsearch/integration/IndexPrivilegeTests.java]: Should be `assertAccessIsDenied`, right?

**Review by albertzaharovits — CHANGES_REQUESTED** [review]
I left a few comments. I think this is close and after you address the comments it should be OK.

**Comment by albertzaharovits** [comment]
Thank you for your contribution @amirhmd and apologies for the delay.
I'll get have another pass after we settle the comments.

**Comment by amirhmd** [comment]
Hi @albertzaharovits  , thanks for checking my code. As soon as I get home I will apply the comment.

File: x-pack/plugin/security/src/test/java/org/elasticsearch/integration/IndexPrivilegeTests.java
Comment (albertzaharovits) [inline x-pack/plugin/security/src/test/java/org/elasticsearch/integration/IndexPrivilegeTests.java]: Please move this to the `testUser11` function and do it for the `b` and `c` indices, i.e.
`assertUserIsDenied("u11", "maintenance", "b");`, `assertUserIsDenied("u11", "maintenance", "c");` .

File: x-pack/docs/en/security/authorization/privileges.asciidoc
Comment (albertzaharovits) [inline x-pack/docs/en/security/authorization/privileges.asciidoc]: You missed an `and`. We try to be sharp wrt the docs.
```suggestion
Permits refresh, flush, synced flush and force merge index administration operations.
```

**Review by albertzaharovits — APPROVED** [review]
I've left two minor comments, please address, otherwise LGTM.
We'll wait for Tim's review and then merge. Thank you for your effort Amir!

**Comment by amirhmd** [comment]
Thanks @albertzaharovits  for the comment. I will commit in 3 hours.

**Comment by amirhmd** [comment]
`elasticsearch-ci/2` is failing because as part of #50454 `Wait For Snapshot` is added. I think thats why this test (`testWaitForSnapshotSlmExecutedBefore`) is failing, is my understanding correct?

**Comment by amirhmd** [comment]
Hi @albertzaharovits may I know if I should do anything in this regard https://github.com/elastic/elasticsearch/pull/50643#issuecomment-573236178?

**Comment by albertzaharovits** [comment]
@amirhmd you don't have to worry about those tests. There's no action to take on your part.
@elasticmachine  update branch
ping @tvernum

**Comment by albertzaharovits** [comment]
@elasticmachine update branch

**Comment by albertzaharovits** [comment]
@elasticmachine update branch

File: x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/security/authz/privilege/IndexPrivilege.java
Comment (tvernum) [inline x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/security/authz/privilege/IndexPrivilege.java]: Minor:
```suggestion
            entry("maintenance", MAINTENANCE));
```

File: x-pack/plugin/security/src/test/java/org/elasticsearch/integration/IndexPrivilegeTests.java
Comment (tvernum) [inline x-pack/plugin/security/src/test/java/org/elasticsearch/integration/IndexPrivilegeTests.java]: I think this this is what we want.
A user beling allowed `maintenance` should not imply anything about whether they are/aren't allowed `crud`.

File: x-pack/plugin/security/src/test/java/org/elasticsearch/integration/IndexPrivilegeTests.java
Comment (tvernum) [inline x-pack/plugin/security/src/test/java/org/elasticsearch/integration/IndexPrivilegeTests.java]: Per above.

**Review by tvernum — CHANGES_REQUESTED** [review]

File: x-pack/plugin/security/src/test/java/org/elasticsearch/integration/IndexPrivilegeTests.java
Comment (amirhmd) [reply x-pack/plugin/security/src/test/java/org/elasticsearch/integration/IndexPrivilegeTests.java]: @tvernum the first and second statement made me a bit confuse. but it looks like the explanation has precedence. so I will remove them.

**Review by tvernum — APPROVED** [review]

**Comment by albertzaharovits** [comment]
Thank you for your contribution, @amirhmd !

**Pull request merged**
